src/pykiso/lib/auxiliaries/udsaux/common/uds_callback.py

```diff
@@ -0,0 +1,225 @@
+
+    :param request: request from the client that should be responded to.
+    :param response: full UDS response to send. If not set, respond with a basic
+        positive response with the specified response_data.
+    :param response_data: UDS data to send. If not set, respond with a basic
+        positive response containing no data.
+    :param data_length: optional length of the data to send in the response if
+        the data has a fixed length (zero-padded).
+    :param callback: custom callback function to register. The callback function
+        must have two positional arguments: the received request as a list of
+        integers and the UdsServerAuxiliary instance.
+    """
+
+    # e.g. 0x1003 or [0x10, 0x03]
+    request: Union[int, List[int]]
+    # e.g. 0x5003 or [0x50, 0x03]
+    response: Optional[Union[int, List[int]]] = None
+    # e.g. 0x1011 or b'DATA'
+    response_data: Optional[Union[int, bytes]] = None
+    # specify zero-padding
+    data_length: Optional[int] = None
+    # custom callback
+    callback: Optional[Callable[[List[int], UdsServerAuxiliary], None]] = None
+
+    def __post_init__(self):
+        """Parse the provided parameters to create a fully formed UDS response."""
+        self.call_count = 0
+
+        if isinstance(self.request, int):
+            self.request = list(UdsCallback.int_to_bytes(self.request))
+
+        if isinstance(self.response, int):
+            self.response = list(UdsCallback.int_to_bytes(self.response))
+        elif not self.response:
+            self.response = [self.request[0] + 0x40] + self.request[1:]
+
+        if self.response_data:
+            self.response_data = (
+                list(UdsCallback.int_to_bytes(self.response_data))
+                if isinstance(self.response_data, int)
+                else list(self.response_data)
+            )
+            self.response.extend(self.response_data)
+
+        if self.data_length is not None and self.response_data:
+            # pad with zeros to reach the expected length
+            self.response.extend([0x00] * (self.data_length - len(self.response_data)))
+
+    def __call__(self, received_request: List[int], aux: UdsServerAuxiliary) -> None:
+        """Trigger the callback and increase the call count.
+
+        If a custom callback function is defined call it, otherwise send
+        the registered response to the client.
+
+        :param received_request: the received UDS request as a list of
+            integers
+        :param aux: the UdsServerAuxiliary instance for which the
+            callback is registered
+        """
+        self.call_count += 1
+        if self.callback is not None:
+            self.callback(aux, received_request)
+        else:
+            aux.send_response(self.response)
+
+    @staticmethod
+    def int_to_bytes(integer: int) -> bytes:
+        return integer.to_bytes((integer.bit_length() + 7) // 8, "big")
+
+
+def handle_data_download(
```

Comment:
I thought this was not specified by the protocol. Did I miss understood or is this a mistake?

Reply:
The DownloadData service is actually standardized by the ISO, so we can leave it here ;)